src/main/java/org/myrobotlab/service/NeoPixel.java

```diff
@@ -26,501 +26,829 @@
-  }
-
-  public void onRegistered(Registration s) {
-    refreshControllers();
-    broadcastState();
+    animationRunner = new AnimationRunner();
+    Double d = fpsToWaitMs.calcOutput(speedFps);
+    animationWaitMs = d.intValue();
   }
 
-  public List<String> refreshControllers() {
-    controllers = Runtime.getServiceNamesFromInterface(NeoPixelController.class);
-    return controllers;
-  }
+  @Override
+  public void attach(Attachable service) throws Exception {
 
-  // @Override
-  public NeoPixelController getController() {
-    return controller;
-  }
+    if (SpeechSynthesis.class.isAssignableFrom(service.getClass())) {
+      attachSpeechSynthesis((SpeechSynthesis) service);
+      return;
+    }
 
-  public String getControllerName() {
-    String controlerName = null;
-    if (controller != null) {
-      controlerName = controller.getName();
+    if (NeoPixelController.class.isAssignableFrom(service.getClass())) {
+      attachNeoPixelController((NeoPixelController) service);
+      return;
     }
-    return controlerName;
+    warn(String.format("%s.attach does not know how to attach to a %s", this.getClass().getSimpleName(), service.getClass().getSimpleName()));
   }
 
-  public boolean isAttached() {
+  @Override
+  public void attachNeoPixelController(NeoPixelController neoCntrlr) {
+
     if (controller != null) {
-      if (((Arduino) controller).getDeviceId((Attachable) this) != null) {
-        isAttached = true;
-        return true;
+      if (controller.equals(neoCntrlr.getName())) {
+        return;
       }
-      controller = null;
+      log.info("{} already attached detach first to attach {}", controller, neoCntrlr.getName());
+      return;
     }
-    isAttached = false;
-    return false;
-  }
 
-  public void setPixel(int address, int red, int green, int blue) {
-    setPixel(address, red, green, blue, 0);
+    if ((pin == null) || (pixelCount == null)) {
+      error("%s pin and pixe count are required before attaching");
+      return;
+    }
+
+    controller = neoCntrlr.getName();
+    neoCntrlr.neoPixelAttach(getName(), pin, pixelCount, pixelDepth);
+    broadcastState();
+  }
+  
+  @Deprecated /* use clear() */
+  public void animationStop() {
+    clear();
+  }
+  
+  @Override
+  public boolean isAttached(Attachable instance) {
+    return instance.getName().equals(controller);
   }
 
-  public void setPixel(int address, int red, int green, int blue, int white) {
-    PixelColor pixel = new PixelColor(address, red, green, blue, white);
-    setPixel(pixel);
+  public void attachSpeechSynthesis(SpeechSynthesis mouth) {
+    subscribe(mouth.getName(), "publishStartSpeaking");
+    subscribe(mouth.getName(), "publishEndSpeaking");
+    // equalizer();
   }
 
-  public void setPixel(String address, String red, String green, String blue) {
-    setPixel(address, red, green, blue, "0");
+  @Override
+  public void clear() {
+    // stop java animations
+    animationRunner.stop();

```

Comment:
this can throw an interrupted exception.. yes?    what if the animation that is running is just on the uc..   we dont' care about the java animation.. the java animation thread shouldn't be running in the first place.

Reply:
thread now is only started when equalizer animation starts